library/java/src/io/envoyproxy/envoymobile/engine/EnvoyHTTPStream.java

```diff
@@ -62,22 +62,29 @@ public void sendTrailers(Map<String, List<String>> trailers) {
     JniLibrary.sendTrailers(streamHandle, toJniLibraryHeaders(trailers));
   }
 
-  /**
-   * Cancel the streamHandle. This functions as an interrupt, and aborts further
-   * callbacks and handling of the streamHandle.
-   *
-   * @return Success, unless the streamHandle has already been canceled.
-   */
-  public int resetStream() { return JniLibrary.resetStream(streamHandle); }
-
   /**
    * Cancel the stream. This functions as an interrupt, and aborts further
    * callbacks and handling of the stream.
+   *
+   * @return int, success unless the stream has already been canceled.
    */
-  public void cancel() { JniLibrary.cancel(); }
+  public int cancel() {
+    // Step 1: atomically and synchronously prevent the execution of further
+    // callbacks other than onCancel.
+    if (!callbacksContext.canceled.getAndSet(true)) {
+      // Step 2: directly fire the cancel callback.
+      callbacksContext.onCancel();
+      // Step 3: propagate the reset into native code.
+      JniLibrary.resetStream(streamHandle);
+      return 0;
+    } else {
+      return 1;
+    }
+  }
```

Comment:
If we delegate this logic to `JvmCallbackContext`, the `AtomicBoolean` can remain private.